_posts/2015-09-01-Cocoa-Architecture:-Dropped-Design-Patterns.markdown

```diff
@@ -0,0 +1,146 @@
+
+It's important to preface that I don't believe in perfect code, or am I a fan of big re-writes. We can spot a bad pattern, but not do anything about it. We do have apps to ship, not a codebase to make perfect for the sake of technical purity.
+
+<!-- more -->
+--------------------------------------------------------------------------------
+
+## NSNotifications as a decoupling method
+
+A lot of the initial codebase for eigen relied on using NSNotifications as a way of passing messages throughout the application. There were notifications for user settings changes, download status updates, anything related to authentication and the corresponding different error states and a few app features. These relied on sending global notifications with very little attempts at scoping the relationship between objects.
+
+NSNotificationCenter notifications are an implementation of the [Observer Pattern](https://en.wikipedia.org/wiki/Observer_pattern) in Cocoa. They are a beginner to intermediate programmers design paradigm dream. It offers a way to have objects send messages to each other without having to go through any real coupling. As someone just starting on iOS ( I had mostly been a OS X programmer before starting at Artsy ) it was an easy choice to adapt.
+
+One of the biggest downsides of using NSNotifications are that they make it easy to be lazy as a programmer. It allows you to not think carefully about the relationships between your objects and instead to pretend that they are loosely coupled, when instead they are coupled but via stringly typed notifications.
+
+Loose-coupling can have it's place but without being careful there is no scope on what could be listening to any notification. Also de-registering for interest can be a tricky thing [to learn](http://stackoverflow.com/questions/tagged/nsnotification) and the default memory-management behavior is about to change ( [for the better](https://developer.apple.com/library/prerelease/mac/releasenotes/Foundation/RN-Foundation/index.html#//apple_ref/doc/uid/TP30000742).)
+
+We still have a [lot of notifications](https://github.com/artsy/energy/blob/702036664a087db218d3aece8ddddb2441f931c8/Classes/Constants/ARNotifications.h) in Energy, however in Eigen and Eidolon there are next to none. We don't even have a specific file for the constants.
+
+## #define kARConstant
+
+Not much to say here, using `#defines` as constants was definitely [in favour](https://github.com/adium/adium/blob/master/Source/AdiumAccounts.m#L24-L30) when I learned Objective-C. Likely a throw back from C. Using `#defines` as constants would not use on-device memory to store an unchanging value, which we used to _really_ care about, but it's likely that a modern copy of LLVM does something similar under the hood. Switching to real variables means you can inspect and use in a debugger and use can rely on the type system better.
```

Comment:
Could be useful to link to or discuss a before/after. "Switching to real variables" makes it sounds like we're using global variables or something, but I think you're talking about extern const's, right?

"would not use on-device memory" not quite sure I understand – could you clarify?